### PR TITLE
Pun labelled arguments with type constraint in function applications

### PR DESCRIPTION
**Pull request by gretay-js**
Punning of labelled parameters with type constraints is already supported in function definitions. This PR proposes to add a similar syntax for function application. For example, labelled argument to a function application can be given as `foo ~(x:int)`  instead of the explicit `foo ~x:(x:int)`.

The motivation comes from writing a ppx, which takes as a payload a function application with labelled arguments. Explicit type constraints was required for this ppx, and I was surprised that punning didn't work in this situation.
There is probably no other compelling use of this syntax, but I think it would be nice to have symmetry between definitions and
applications, and the additional syntax doesn't seem to cause any parser conflicts.

By the way, parser/HACKING.adoc mentions `make build-all-asts` for testing but this make target does not exist any more.
I would like to add a small test for the new syntax. Where in the testsuite should it go?

### CODE REVIEW

**Comment by nojb** [comment]
> By the way, parser/HACKING.adoc mentions `make build-all-asts` for testing but this make target does not exist any more.

This target is in the `Makefile.dev` file at the root of the repository.

> I would like to add a small test for the new syntax. Where in the testsuite should it go?

You could add it to `testsuite/tests/parsetree/locations_test.ml` or perhaps to `testsuite/tests/parsetree/source.ml` (or to both). The first test just dumps the AST (using `-dparsetree`) and compares to the reference file. The second file is parsed, printed out (using `Pprintast`) and reparsed a second time and it is checked that the same AST is obtained.

By the way, is the bootstrap necessary?

**Comment by garrigue** [comment]
Could you expand a bit on why this would be useful for ppx's ?
My understanding is that punning of labelled variables is purely syntactic sugar, and is already expanded in the syntax tree. So I don't see how adding the same superficial syntactic sugar for expressions would help with program transformations.
I agree that symmetry is in general a good property, but in this case this syntax in the context of expressions leaves me scratching my head. Wouldn't it be more natural to add the type annotation where you define the variable, rather than where you use it? This asymmetry between definition and use is precisely why this punning is useful for function definitions.
Actually, the only case where it seems to make sense (semantically) is if you where to use exactly the same type as in the definition, but then there is no need to annotate.

**Comment by lpw25** [comment]
> My understanding is that punning of labelled variables is purely syntactic sugar, and is already expanded in the syntax tree. So I don't see how adding the same superficial syntactic sugar for expressions would help with program transformations.

It is useful as the input to a ppx, rather than in the generated code. e.g. you can have a ppx where people write:
```ocaml
[%foo bar ~(x : int) ~(y : float) ~z:(expr : string)]
```

For instance, this would be a slight improvement for [ppx_sexp_message](https://github.com/janestreet/ppx_sexp_message), which currently supports the following forms:
```ocaml
[%message "error" ~tag:(msg : string)]
[%message "error" ~_:(msg : string)]
[%message "error" (msg : string)] 
[%message "error" "value of msg"]
```
Note that the `~` currently has to be dropped for the punned version. With this PR it would not need to be.

**Comment by gretay-js** [comment]
@nojb thank you for your help. 

> This target is in the `Makefile.dev` file at the root of the repository.

Found it, thanks! All ast tests pass. 

> You could add it to `testsuite/tests/parsetree/locations_test.ml` or perhaps to `testsuite/tests/parsetree/source.ml` (or to both). The first test just dumps the AST (using `-dparsetree`) and compares to the reference file. The second file is parsed, printed out (using `Pprintast`) and reparsed a second time and it is checked that the same AST is obtained.

I added a test to both files.
 
> By the way, is the bootstrap necessary?

I was not sure. It seems to work fine without the bootstrap, so I've removed that commit.

**Comment by gretay-js** [comment]
> Could you expand a bit on why this would be useful for ppx's ?
 @lpw25 beat me to it!

**Comment by gretay-js** [comment]
I've rebase this PR. Could someone review this PR please? The new syntax does not change the parsetree.

File: parsing/parser.mly [hunk 1]
```diff
@@ -2467,6 +2467,9 @@ labeled_simple_expr:
   | TILDE label = LIDENT
       { let loc = $loc(label) in
         (Labelled label, mkexpvar ~loc label) }
+  | TILDE LPAREN label = LIDENT ty = type_constraint RPAREN
+      { let loc = $loc(label) in
+        (Labelled label, mkexp_constraint ~loc (mkexpvar ~loc label) ty) }
```
Comment (nchataing) [inline parsing/parser.mly]: I wouldn't put the location of the label for the constraint expression, but `~loc:($startpos($2), $endpos)` instead.

For reference, the parse tree of `(0 : int)` (in the toplevel) is : 
```
Ptop_def
  [
    structure_item (//toplevel//[1,0+0]..[1,0+9]) ghost
      Pstr_eval
      expression (//toplevel//[1,0+0]..[1,0+9]) ghost
        Pexp_constraint
        expression (//toplevel//[1,0+1]..[1,0+2])
          Pexp_constant PConst_int (1,None)
        core_type (//toplevel//[1,0+5]..[1,0+8])
          Ptyp_constr "int" (//toplevel//[1,0+5]..[1,0+8])
          []
  ]
```
where the constraint location encapsulates everything.

**Review by nchataing — CHANGES_REQUESTED** [review]
Looks good to me except for the location of the constraint expression :-)

File: parsing/parser.mly (hunk 1 above)
Comment (gretay-js) [reply parsing/parser.mly]: Fixed, thank you.

**Review by nchataing — APPROVED** [review]
Perfect!

**Review by gasche — APPROVED** [review]
Approved on @nchataing's behalf.

**Pull request merged**